shared/metrics/src/main/java/io/pravega/shared/metrics/StatsProviderImpl.java

```diff
@@ -66,84 +52,93 @@ public void start() {
+                    return Duration.ofSeconds(conf.getStatsOutputFrequencySeconds().getSeconds());
+                }
+
+                @Override
+                public String prefix() {
+                    return conf.getMetricsPrefix();
+                }
+
+                @Override
+                public String host() {
+                    return conf.getStatsDHost();
+                }
+
+                @Override
+                public int port() {
+                    return conf.getStatsDPort();
+                }
+
+                @Override
+                public StatsdFlavor flavor() {
+                    return StatsdFlavor.TELEGRAF; //Dimension supported
+                }
+
+                @Override
+                public String get(String key) {
+                    return null;
+                }
+            };
+
+            MeterRegistry registry = new StatsdMeterRegistry(config, Clock.SYSTEM);
+            Metrics.addRegistry(registry);
         }
-        for (ScheduledReporter r : reporters) {
-            r.start(conf.getStatsOutputFrequencySeconds().getSeconds(), TimeUnit.SECONDS);
+
+        if (conf.isEnableInfluxDBReporter()) {
+            log.info("Configuring stats with direct InfluxDB at {}", conf.getInfluxDBUri());
+            InfluxConfig config = new InfluxConfig() {
+                @Override
+                public Duration step() {
+                    return Duration.ofSeconds(conf.getStatsOutputFrequencySeconds().getSeconds());
+                }
+                
+                @Override
+                public String prefix() {
+                    return conf.getMetricsPrefix();
+                }
+
+                @Override
+                public String uri() {
+                    return conf.getInfluxDBUri();
+                }
+
+                @Override
+                public String db() {
+                    return conf.getInfluxDBName();
+                }
+
+                @Override
+                public String userName() {
+                    return conf.getInfluxDBUserName();
+                }
+
+                @Override
+                public String password() {
+                    return conf.getInfluxDBPassword();
+                }
+
+                @Override
+                public String get(String k) {
+                    return null;
+                }
+            };
+
+            MeterRegistry registry = new InfluxMeterRegistry(config, Clock.SYSTEM);
+            Metrics.addRegistry(registry);
         }
+        //TODO: add more registries
     }
 
     @Synchronized
     @Override
     public void close() {
-        for (ScheduledReporter r : reporters) {
-            try {
-                r.report();
-                r.stop();
-            } catch (Exception e) {
-                log.error("Exception report or stop reporter", e);
-            }
-        }
-
-        metrics.removeMatching(MetricFilter.ALL);
+        metrics.close();
```

Comment:
Why do you close `metrics`? You do no own this instance; it's a reference to an external, static field!

In `close` you only need to clean up your own resources, not external resources. For example, you should remove the recorded registries that you added.

Reply:
This is to close registry so it stops accepting new metric object, also stops sending more metrics.
Micrometer registry holds weak references to metric object, so there's no need to explicitly remove all metric objects inside.

Reply:
The point is that you are closing an object that you do not own. This is a rather bad design and should be avoided as it may have unintended side effects.

Either create a new such instance with your class or don't close that object to which you have a static reference.

Reply:
Fixed by removing all concrete registry inside the composite registry.
I think the registry.close() is the api for application to manage the lifecycle of registry. For example, when we start and close segmentstore, we start and close metric registry accordingly.